driver-core/src/main/com/mongodb/internal/connection/BsonWriterHelper.java

```diff
@@ -65,16 +96,86 @@ static void writePayload(final BsonWriter writer, final BsonOutput bsonOutput, f
         }
 
         if (payload.getPosition() == 0) {
-            throw new BsonMaximumSizeExceededException(format("Payload document size is larger than maximum of %d.",
-                    payloadSettings.getMaxDocumentSize()));
+            throw createBsonMaximumSizeExceededException(payloadSettings.getMaxDocumentSize());
         }
     }
 
+    /**
+     * @return See {@link DualMessageSequences#encodeDocuments(WritersProviderAndLimitsChecker)}.
+     */
+    static EncodeDocumentsResult writeDocumentsOfDualMessageSequences(
+            final DualMessageSequences dualMessageSequences,
+            final int commandDocumentSizeInBytes,
+            final BsonOutput firstOutput,
+            final BsonOutput secondOutput,
+            final MessageSettings messageSettings) {
+        BsonBinaryWriter firstWriter = createBsonBinaryWriter(firstOutput, dualMessageSequences.getFirstFieldNameValidator(), null);
+        BsonBinaryWriter secondWriter = createBsonBinaryWriter(secondOutput, dualMessageSequences.getSecondFieldNameValidator(), null);
+        // the size of operation-agnostic command fields (a.k.a. extra elements) is counted towards `messageOverheadInBytes`
+        int messageOverheadInBytes = 1000;
```

Comment:
I am not sure if limitation is the right word here. But yes, this is unique to client bulk writes:

1. A mixed bulk write does not mix different kinds of operations in a batch because it is technically impossible there. As a result, we know in advance what kind of operations is in a batch, thus knowing whether it supports retries or not. That allows us to encode `extraElements` before encoding the `PAYLOAD_TYPE_1_DOCUMENT_SEQUENCE` section. Thus, when we are encoding the document sequence, we know exactly how much space we have left available before reaching `MessageSettings.getMaxMessageSize`.
2. A client bulk write may mix different kinds of operations in a batch, which means that we know whether the batch supports retries only after encoding its `PAYLOAD_TYPE_1_DOCUMENT_SEQUENCE` sections. That is, we may need to write something after writing those sections. That, in turn, means we can't know exactly how much space we have left when we encode document sequences. But whatever we write after writing the sequences, its size is bounded, and [1000 bytes is used in the spec as the value that is definitely not smaller than that bound](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#total-message-size).